database.cc

```diff
@@ -1401,6 +1401,9 @@ compare_atomic_cell_for_merge(atomic_cell_view left, atomic_cell_view right) {
 future<std::tuple<lw_shared_ptr<query::result>, cache_temperature>>
 database::query(schema_ptr s, const query::read_command& cmd, query::result_options opts, const dht::partition_range_vector& ranges,
                 tracing::trace_state_ptr trace_state, db::timeout_clock::time_point timeout) {
+    if (cmd.slice.options.contains(query::partition_slice::option::reversed)) {
+        s = s->make_reversed();
```

Comment:
This operation s->make_reversed(), unless I'm misunderstanding something, is fairly elaborate and slow. Should we perhaps **cache** it result? Most easily, the schema s itself could cache a schema_ptr of its reversed version and return it when asked for it a second time.
(to avoid a shardptr loop and memory leak, the reverse will need to hold a weak pointer to the original, not a regular shared pointer...).

Reply:
This happens exactly once per query. Are you suggesting to cache this across queries? We can use the schema registry for that by making it aware of the flipped UUID bits (BTW what do you as our foremost expert on UUIDs think?) and doing the reverse if the reversed schema is not present in the registry. I think this can be left for a follow-up. We are not aiming for top-notch performance in the initial version.

Reply:
This can be done as a follow-up, but I also think we should cache it, it may be non-negligible if reads are fast.

Reply:
It's okay for a schema object to cache its reversed counterpart in a mutable member.

Reply:
Another alternative is for the schema to create its reversed counterpart when it's built the first time.

Reply:
I think using the schema registry for this will have better results -- it will cover crossing shards too.

Reply:
As a bonus: once a schema is reversed, any reversing will yield the same C++ object on that shard (as long as the registry entry is live), so we won't have a proliferation of schema objects for the same schema version.

Reply:
This includes double reversing: `s->make_reversed()->make_reversed() == s` will be true.